### PR TITLE
update metadata

### PR DESCRIPTION
**Pull request by jiangxinmeng1**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/vm/engine/tae/common/dlnode.go [hunk 1]
```diff
@@ -208,9 +208,13 @@ func (it *LinkIt) Valid() bool {
 }
 
 func (it *LinkIt) Next() {
-	it.linkLocker.RLock()
-	it.curr = it.nextFunc(it.curr)
-	it.linkLocker.RUnlock()
+	if it.linkLocker != nil {
```
Comment (fengttt) [inline pkg/vm/engine/tae/common/dlnode.go:211]: There are so many questions regarding this little library.

Who is this lock guarding?  Obviously INSERT or DELETE to the list does not acquire this lock.   So, what is the point?

File: pkg/vm/engine/tae/catalog/block.go [hunk 2]
```diff
@@ -160,22 +138,23 @@ func (entry *BlockEntry) GetFileTs() (types.TS, error) {
...
 }
 func (entry *BlockEntry) PrepareRollback() (err error) {
-	entry.RLock()
-	currOp := entry.CurrOp
-	entry.RUnlock()
-	if currOp == OpCreate {
+	entry.Lock()
+	err = entry.BaseEntry.PrepareRollbackLocked()
+	if err != nil {
+		panic(err)
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/block.go:144]: Isn't here a deadlock?   You need to defer?

File: pkg/vm/engine/tae/catalog/block.go [hunk 3]
```diff
@@ -160,22 +138,23 @@ func (entry *BlockEntry) GetFileTs() (types.TS, error) {
...
+	}
+	empty := entry.IsEmpty()
+	entry.Unlock()
+	if empty {
 		if err = entry.GetSegment().RemoveEntry(entry); err != nil {
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/block.go:149]: And this need to get a lock?

File: pkg/vm/engine/tae/catalog/catalog_test.go [hunk 4]
```diff
@@ -369,101 +368,101 @@ func TestTable1(t *testing.T) {
...
-	r = bytes.NewBuffer(buf)
-
-	cmd, _, err = txnbase.BuildCommandFrom(r)
-	assert.Nil(t, err)
-	eCmd = cmd.(*EntryCommand)
-	assert.Equal(t, tb.ID, eCmd.entry.ID)
-	assert.Equal(t, tb.DeleteAt, eCmd.entry.DeleteAt)
-	assert.Equal(t, tb.db.ID, eCmd.DBID)
-}
+// func TestCommand(t *testing.T) {
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/catalog_test.go:371]: Why do you need to disable this test?

File: pkg/vm/engine/tae/catalog/checkpoint.go [hunk 5]
```diff
@@ -47,86 +47,29 @@ func CheckpointSelectOp(entry *BaseEntry, minTs, maxTs types.TS) bool {
...
-		ckpEntry.AddCommand(cloned.MakeLogEntry())
-		return
-	}
-	// 4.2 entry was not deleted
-	if !entry.HasDropped() {
-		ckpEntry.AddIndex(entry.LogIndex)
-		cloned := item.Clone()
-		entry.RUnlock()
-		ckpEntry.AddCommand(cloned.MakeLogEntry())
+	ckpItem := entry.GetCheckpointItems(minTs, maxTs)
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/checkpoint.go:66]: Again, here the lock/unlock and what the lock protects, are all very confusing.    I don't think it is right.

File: pkg/vm/engine/tae/catalog/command.go [hunk 6]
```diff
@@ -351,6 +297,9 @@ func (cmd *EntryCommand) WriteTo(w io.Writer) (n int64, err error) {
 			return
 		}
 		sn, err = cmd.Block.WriteTo(w)
+		if err != nil {
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/command.go:300]: It this, all about Serialize and DeSerialize?    Should you consider get rid of all these code and use protocol buffer?

File: pkg/vm/engine/tae/catalog/database.go [hunk 7]
```diff
@@ -232,6 +219,12 @@ func (e *DBEntry) DropTableEntry(name string, txnCtx txnif.AsyncTxn) (deleted *T
 	entry := dn.GetPayload().(*TableEntry)
 	entry.Lock()
 	defer entry.Unlock()
+	needWait, txn := entry.NeedWaitCommitting(txnCtx.GetStartTS())
+	if needWait {
+		entry.Unlock()
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/database.go:224]: Lock again?

**Review by fengttt — CHANGES_REQUESTED** [review]
This PR has too many problems.

1. Why do you write your own ser/deser instead of protocol buffer.
2. More importantly, I don't think we designed, or understood all these locking.   XuPeng need to review the lock design and maybe completely redo locking.

**Comment by fengttt** [comment]
Had a long discussion off line with Xu Peng.   I will let him to decide on the PR.   

Thank you,

**Review by XuPeng-SH — APPROVED** [review]
Some comments were added, and the names of some interfaces were modified. Have run bvt offline several times and passed the stability test for a few hours. According to the current implementation, there should be no problem.

**Comment by XuPeng-SH** [comment]
> Had a long discussion off line with Xu Peng. I will let him to decide on the PR.
> 
> Thank you,

Rename `Link` to `SortedDList` and add some comments.

Here's how engine manage all locks:
image

Have run bvt offline many times and passed the stability test for a few hours.

File: pkg/vm/engine/tae/catalog/base.go [hunk 8]
```diff
@@ -22,466 +22,569 @@ import (
...
-	n = sn + 8 + 8 + 8
+	n += 8
+	be.MVCC.Loop(func(node *common.DLNode) bool {
+		var n2 int64
+		n2, err = node.GetPayload().(*UpdateNode).WriteTo(w)
+		if err != nil {
+			return false
+		}
+		n += n2
+		return true
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/base.go:327]: Is this a closure?

Comment (jiangxinmeng1) [reply pkg/vm/engine/tae/catalog/base.go:327]: No. When return true, it goes next. When return false, it stops.

Comment (jiangxinmeng1) [reply pkg/vm/engine/tae/catalog/base.go:327]: Yes, it's a closure.

File: pkg/vm/engine/tae/catalog/base.go [hunk 9]
```diff
@@ -22,466 +22,569 @@ import (
...
+		}
+		be.MVCC.Insert(un)
+		n += n2
+	}
+	return
 }
 
-func (be *BaseEntry) GetID() uint64 { return be.ID }
-
 func (be *BaseEntry) DoCompre(oe *BaseEntry) int {
```
Comment (fengttt) [inline pkg/vm/engine/tae/catalog/base.go:369]: Now you introduced an compare on be.id and oe.id -- does this ordering based on id, actually always work?  I don't know, but ask Xu Peng to confirm.

Will the id of be and oe ever change?      If not, why this rlock/runlock is useful?

**Review by fengttt — COMMENTED** [review]
Overall, this is tricky, delicate code.   The only way to make it strong enough is massive concurrency test.

I don't think here the test cases are anywhere close to enough for people to get enough confidence.     We need to have a discussion on testing.

**Review by fengttt — APPROVED** [review]

**Pull request merged**